File: gopro_python_bridge.py
#!/usr/bin/env python3
"""
GoPro Python Bridge - Uses official Open GoPro SDK for reliable camera control
Communicates with JavaScript app via JSON over stdin/stdout
"""

import asyncio
import json
import sys
import logging
from open_gopro import GoPro

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class GoProBridge:
    def __init__(self):
        self.gopro = None
        self.is_connected = False

    async def connect(self):
        """Connect to GoPro using BLE only"""
        try:
            logger.info("Connecting to GoPro via BLE only...")
            self.gopro = GoPro(enable_wifi=False)  # Disable WiFi, BLE only

            logger.info("Opening GoPro connection...")
            self.gopro.open()  # Not async
            logger.info("GoPro.open() completed")

            # Check if camera is ready
            logger.info("Checking BLE connection status...")
            if not self.gopro.is_ble_connected:
                logger.warning("BLE connection check failed")
                raise Exception("BLE connection failed")

            self.is_connected = True
            logger.info("Successfully connected to GoPro via BLE")
            return {
                "success": True,
                "message": "Connected to GoPro via BLE",
                "connected": True,
                "ble_connected": True,
            }

        except Exception as e:
            logger.error(f"BLE connection failed: {e}")
            self.is_connected = False
            return {"success": False, "error": str(e), "connected": False}

    async def disconnect(self):
        """Disconnect from GoPro"""
        try:
            if self.gopro:
                self.gopro.close()  # Not async
            self.is_connected = False
            logger.info("Disconnected from GoPro")
            return {
                "success": True,
                "message": "Disconnected from GoPro",
                "connected": False,
            }
        except Exception as e:
            logger.error(f"Disconnect error: {e}")
            self.is_connected = False
            return {"success": False, "error": str(e), "connected": False}

    async def get_status(self):
        """Get camera status using BLE commands"""
        try:
            if not self.is_connected or not self.gopro:
                return {"success": False, "error": "Not connected"}

            # Get status using BLE status properties
            busy = self.gopro.ble_status.system_busy
            encoding = self.gopro.ble_status.encoding_active
            ready = self.gopro.ble_status.system_ready

            return {
                "success": True,
                "status": {
                    "busy": str(busy),
                    "encoding": str(encoding),
                    "ready": str(ready),
                    "group": 1,  # Default group
                },
            }
        except Exception as e:
            logger.error(f"BLE status error: {e}")
            return {"success": False, "error": str(e)}

    async def check_connection(self):
        """Check if GoPro is still connected"""
        try:
            if not self.gopro:
                return {
                    "success": True,
                    "connected": False,
                    "ble_connected": False,
                    "message": "No GoPro instance",
                }

            # Check BLE connection status
            ble_connected = self.gopro.is_ble_connected
            self.is_connected = ble_connected

            return {
                "success": True,
                "connected": ble_connected,
                "ble_connected": ble_connected,
                "message": "Connected" if ble_connected else "Disconnected",
            }
        except Exception as e:
            logger.error(f"Connection check error: {e}")
            self.is_connected = False
            return {
                "success": False,
                "error": str(e),
                "connected": False,
                "ble_connected": False,
            }

    async def take_photo(self):
        """Take photo using BLE commands"""
        try:
            if not self.is_connected or not self.gopro:
                return {"success": False, "error": "Not connected"}

            logger.info("Taking photo using BLE commands...")

            # Use BLE command to take photo
            result = self.gopro.ble_command.set_shutter(1)

            # Handle response
            if hasattr(result, "status"):
                status = result.status
            else:
                status = "SUCCESS"  # Assume success if no status field

            if status == "SUCCESS" or str(status) == "ErrorCode.SUCCESS":
                logger.info("Photo taken successfully via BLE!")
                return {"success": True, "message": "Photo taken successfully via BLE"}
            else:
                return {"success": False, "error": f"Photo failed: {status}"}

        except Exception as e:
            logger.error(f"BLE photo error: {e}")
            return {"success": False, "error": str(e)}

    async def handle_command(self, command_data):
        """Handle commands from JavaScript app"""
        command = command_data.get("command")
        command_id = command_data.get("commandId")

        result = None
        if command == "connect":
            result = await self.connect()
        elif command == "disconnect":
            result = await self.disconnect()
        elif command == "status":
            result = await self.get_status()
        elif command == "check_connection":
            result = await self.check_connection()
        elif command == "take_photo":
            result = await self.take_photo()
        else:
            result = {"success": False, "error": f"Unknown command: {command}"}

        # Include commandId in response
        if command_id is not None:
            result["commandId"] = command_id

        return result


async def main():
    """Main loop - read JSON commands from stdin, send responses to stdout"""
    bridge = GoProBridge()

    logger.info("GoPro Python Bridge started - waiting for commands...")

    try:
        while True:
            # Read command from stdin
            line = sys.stdin.readline()
            if not line:
                break

            logger.info(f"Received command: {line.strip()}")

            try:
                command_data = json.loads(line.strip())
                logger.info(f"Parsed command: {command_data}")
                result = await bridge.handle_command(command_data)
                logger.info(f"Command result: {result}")

                # Send response to stdout
                print(json.dumps(result))
                sys.stdout.flush()

            except json.JSONDecodeError as e:
                logger.error(f"JSON decode error: {e}")
                error_result = {"success": False, "error": f"Invalid JSON: {e}"}
                print(json.dumps(error_result))
                sys.stdout.flush()

    except KeyboardInterrupt:
        logger.info("Shutting down...")
        await bridge.disconnect()
    except Exception as e:
        logger.error(f"Unexpected error: {e}")
        await bridge.disconnect()


if __name__ == "__main__":
    asyncio.run(main())

File: goproPythonBridge.js
// goproPythonBridge.js
// JavaScript wrapper for GoPro Python Bridge using official Open GoPro SDK

const { spawn } = require('child_process');
const EventEmitter = require('events');

class GoProPythonBridge extends EventEmitter {
  constructor() {
    super();
    this.pythonProcess = null;
    this.isConnected = false;
    this.commandQueue = [];
    this.pendingCommands = new Map();
    this.commandId = 0;
  }

  start() {
    if (this.pythonProcess) {
      this.log('Python bridge already running', 'WARN');
      return;
    }

    this.log('Starting GoPro Python Bridge...', 'INFO');

    // Start Python bridge process using virtual environment
    const pythonPath = process.platform === 'win32' ? 'venv/Scripts/python.exe' : 'venv/bin/python';
    this.pythonProcess = spawn(pythonPath, ['gopro_python_bridge.py'], {
      cwd: process.cwd(),
      stdio: ['pipe', 'pipe', 'pipe']
    });

    // Handle stdout (responses from Python)
    this.pythonProcess.stdout.on('data', (data) => {
      const lines = data
        .toString()
        .split('\n')
        .filter((line) => line.trim());
      lines.forEach((line) => {
        try {
          const response = JSON.parse(line);
          this.handleResponse(response);
        } catch (e) {
          this.log(`Failed to parse Python response: ${line}`, 'ERROR');
        }
      });
    });

    // Handle stderr (Python errors)
    this.pythonProcess.stderr.on('data', (data) => {
      const message = data.toString().trim();
      // Don't treat INFO messages as errors
      if (message.includes('INFO:') || message.includes('GoPro Python Bridge started')) {
        this.log(message, 'INFO');
      } else {
        this.log(`Python stderr: ${message}`, 'ERROR');
      }
    });

    // Handle process exit
    this.pythonProcess.on('exit', (code) => {
      this.log(`Python bridge exited with code ${code}`, 'WARN');
      this.pythonProcess = null;
      this.isConnected = false;
      this.emit('disconnected');
    });

    this.log('GoPro Python Bridge started', 'SUCCESS');
  }

  stop() {
    if (this.pythonProcess) {
      this.log('Stopping GoPro Python Bridge...', 'INFO');
      this.pythonProcess.kill();
      this.pythonProcess = null;
      this.isConnected = false;
    }
  }

  async sendCommand(command, data = {}) {
    return new Promise((resolve, reject) => {
      if (!this.pythonProcess) {
        reject(new Error('Python bridge not running'));
        return;
      }

      const commandId = ++this.commandId;
      const commandData = {
        id: commandId,
        command: command,
        ...data
      };

      // Store pending command
      this.pendingCommands.set(commandId, { resolve, reject });

      // Send command to Python
      this.pythonProcess.stdin.write(JSON.stringify(commandData) + '\n');

      // Set timeout
      setTimeout(() => {
        if (this.pendingCommands.has(commandId)) {
          this.pendingCommands.delete(commandId);
          reject(new Error(`Command timeout: ${command}`));
        }
      }, 10000); // 10 second timeout
    });
  }

  handleResponse(response) {
    const { id, success, error, message, status } = response;

    // Update connection status based on response
    if (error === 'Not connected' && this.isConnected) {
      this.isConnected = false;
      this.emit('disconnected');
    }

    if (id && this.pendingCommands.has(id)) {
      const { resolve, reject } = this.pendingCommands.get(id);
      this.pendingCommands.delete(id);

      if (success) {
        resolve({ message, status });
      } else {
        reject(new Error(error || 'Unknown error'));
      }
    } else {
      // Handle unsolicited responses - these are likely responses without IDs
      if (success !== undefined) {
        // This is a valid response, find the oldest pending command
        const oldestCommand = Array.from(this.pendingCommands.entries())[0];
        if (oldestCommand) {
          const [commandId, { resolve, reject }] = oldestCommand;
          this.pendingCommands.delete(commandId);

          if (success) {
            resolve({ message, status });
          } else {
            reject(new Error(error || 'Unknown error'));
          }
        } else {
          this.log(`Unsolicited response: ${JSON.stringify(response)}`, 'INFO');
        }
      } else {
        this.log(`Unsolicited response: ${JSON.stringify(response)}`, 'INFO');
      }
    }
  }

  // GoPro control methods
  async connect() {
    try {
      this.log('Connecting to GoPro via Python bridge...', 'INFO');
      const result = await this.sendCommand('connect');
      this.isConnected = true;
      this.log('Connected to GoPro successfully', 'SUCCESS');
      this.emit('connected');
      return result;
    } catch (e) {
      this.log(`Connection failed: ${e.message}`, 'ERROR');
      throw e;
    }
  }

  async disconnect() {
    try {
      this.log('Disconnecting from GoPro...', 'INFO');
      const result = await this.sendCommand('disconnect');
      this.isConnected = false;
      this.log('Disconnected from GoPro', 'SUCCESS');
      this.emit('disconnected');
      return result;
    } catch (e) {
      this.log(`Disconnect failed: ${e.message}`, 'ERROR');
      throw e;
    }
  }

  async getCameraStatus() {
    try {
      const result = await this.sendCommand('status');
      return result.status || { busy: 0, encoding: 0, ready: 1, group: 1 };
    } catch (e) {
      this.log(`Status query failed: ${e.message}`, 'ERROR');
      return { busy: undefined, encoding: undefined, ready: undefined, group: undefined };
    }
  }

  async snapPhoto(maxRetries = 3) {
    this.log('===== PHOTO CAPTURE START (Python SDK) =====', 'INFO');

    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        this.log(`Photo attempt ${attempt}/${maxRetries}`, 'INFO');

        const result = await this.sendCommand('take_photo');
        this.log(`Photo captured successfully on attempt ${attempt}`, 'SUCCESS');
        this.log('===== PHOTO CAPTURE END =====', 'INFO');
        return true;
      } catch (e) {
        this.log(`Photo attempt ${attempt} failed: ${e.message}`, 'ERROR');

        if (attempt < maxRetries) {
          const waitTime = 1000;
          this.log(`Waiting ${waitTime}ms before retry...`, 'INFO');
          await new Promise((resolve) => setTimeout(resolve, waitTime));
        } else {
          throw new Error(`All photo attempts failed. Last error: ${e.message}`);
        }
      }
    }

    this.log('===== PHOTO CAPTURE END (FAILED) =====', 'ERROR');
    return false;
  }

  ready() {
    return this.pythonProcess !== null && this.isConnected;
  }

  log(message, level = 'INFO') {
    this.emit('log', `[${level}] ${message}`);
  }
}

module.exports = GoProPythonBridge;
